src/Server/TCPHandler.cpp

```diff
@@ -873,17 +875,71 @@ void TCPHandler::receiveQuery()
         client_info.initial_query_id = client_info.current_query_id;
         client_info.initial_address = client_info.current_address;
     }
-    else
-    {
-        query_context->setInitialRowPolicy();
-    }
 
     /// Per query settings are also passed via TCP.
     /// We need to check them before applying due to they can violate the settings constraints.
     auto settings_format = (client_revision >= DBMS_MIN_REVISION_WITH_SETTINGS_SERIALIZED_AS_STRINGS) ? SettingsWriteFormat::STRINGS_WITH_FLAGS
                                                                                                       : SettingsWriteFormat::BINARY;
     Settings passed_settings;
     passed_settings.read(*in, settings_format);
+
+    /// Interserver secret.
+    bool interserver_secret = client_revision >= DBMS_MIN_REVISION_WITH_INTERSERVER_SECRET;
+    std::string received_cluster;
+    std::string received_hash;
+    if (interserver_secret)
+    {
+        readStringBinary(received_cluster, *in);
+        readStringBinary(received_hash, *in, 32);
+    }
+
+    readVarUInt(stage, *in);
+    state.stage = QueryProcessingStage::Enum(stage);
+
+    readVarUInt(compression, *in);
+    state.compression = static_cast<Protocol::Compression>(compression);
+
+    readStringBinary(state.query, *in);
+
+    /// It is OK to check only when query != INITIAL_QUERY,
+    /// since only in that case the actions will be done.
+    if (interserver_secret && !received_cluster.empty() && client_info.query_kind != ClientInfo::QueryKind::INITIAL_QUERY)
+    {
+#if USE_SSL
+        std::string cluster_secret(query_context->getCluster(received_cluster)->getSecret());
+        std::string data(cluster_secret);
+        data += state.query;
+        data += state.query_id;
+        data += connection_context.getUserName();
+        data += client_info.initial_user;
+
+        std::string calculated_hash;
+        calculated_hash.resize(32);
+
+        if (received_hash.size() != calculated_hash.size())
+            throw NetException("Unexpected hash received from client", ErrorCodes::UNEXPECTED_PACKET_FROM_CLIENT);
+
+        encodeSHA256(data, reinterpret_cast<unsigned char *>(calculated_hash.data()));
+
+        if (calculated_hash != received_hash)
+            throw NetException("Hash mismatch", ErrorCodes::UNEXPECTED_PACKET_FROM_CLIENT);
+        /// TODO: change error code?
+
+        query_context->setUserSecure(client_info.initial_user, socket().peerAddress());
```

Comment:
So you're going to call `setUserSecure()` for each query. I don't think it's unnecessary. I suppose it's probably better to call this function once per connection in `receiveHello()` and just check hash in `receiveQuery()`.

Reply:
But the initial_user may be different for each query, `ClientInfo` (that contains `initial_user`) is passed to the `Connection::sendQuery` and then on the other side of socket `TCPHandler::receiveQuery` it is received and interpreted.

Reply:
ok.